### PR TITLE
Trigger to generate Windows boot files

### PR DESCRIPTION
**Pull request by tpw56j**
Dependencies: python3-pefile python3-hivex ntfs-3g fuse wimlib-utils

Additional metadata
distro kernel-options: post_install - post-installation script
profile kernel-options: pxeboot bootmgr bcd winpe sif
distro_signatures.json: windows": {"2003": {},"2008": {},"2012": {},"2016": {},"XP": {},"7": {},"8": {},"10": {}}

### CODE REVIEW

**Comment by tpw56j** [comment]
When implementing the trigger, a problem arose: an incorrect BSD file was created through python3-hivex. Therefore, the python implementation of bcdedit has been commented out and replaced with hivexsh.

**Comment by dansou901** [comment]
There is one thing I can ask. What about Windows Server 2019?

**Comment by tpw56j** [comment]
I have not yet tried installing Windows Server 2019, but I will definitely try it in the near future.

**Comment by SchoolGuy** [comment]
@tpw56j Would you say this even covers all points of https://github.com/cobbler/cobbler/issues/1342 ?

**Comment by tpw56j** [comment]
> @tpw56j Would you say this even covers all points of #1342 ?

No, in this case we don't need any tools from Microsoft (SSCM). In addition, the trigger does not make the necessary changes to the tftp server configuration.

**Comment by tpw56j** [comment]
> There is one thing I can ask. What about Windows Server 2019?

I added information about Windows Server 2019 to trigger code and distro_signatures.json. The installation was successful.

**Comment by dansou901** [comment]
Thanks for that! Also here, I'll add my review later when I have time.

File: docs/user-guide/windows.rst
Comment (dansou901) [inline docs/user-guide/windows.rst:7]: without "does"

File: docs/user-guide/windows.rst
Comment (dansou901) [inline docs/user-guide/windows.rst:8]: string instead of sting

File: docs/user-guide/windows.rst
Comment (dansou901) [inline docs/user-guide/windows.rst:10]: pxelinux instead of pxelibux

**Review by dansou901 — CHANGES_REQUESTED** [review]
see inline comments

File: docs/user-guide/windows.rst
Comment (tpw56j) [reply docs/user-guide/windows.rst]: @dansou901 Thanks! In commit 0537040, I fixed all these grammar mistakes.

**Review by dansou901 — APPROVED** [review]
LGTM now

**Comment by dansou901** [comment]
But, there are still a lot of whitespaces in the trigger. I guess that's something which should be fixed before merging.

**Comment by SchoolGuy** [comment]
@tpw56j Unrelated to this PR: Can we validate the validity of the autoinstall xml file? Or can we only say that this is valid xml code?

**Review by SchoolGuy — CHANGES_REQUESTED** [review]
Please fix the linter. The CI is our most important way of ensuring that Cobbler works. Since the linter is already complaining, please shush it by appliying the PEP codestyle.

You can execute the checks locally through `make qa` in the project root if the linters are installed on your system.

**Comment by SchoolGuy** [comment]
Again! Thanks a lot for the contribution! The blogpost is already on the way to the website and thus we will merge this!

**Comment by tpw56j** [comment]
> @tpw56j Unrelated to this PR: Can we validate the validity of the autoinstall xml file? Or can we only say that this is valid xml code?

@SchoolGuy autoinstall xml is easiest to create using Microsoft tools (Windows System Image Manager from the Windows ADK). There are also tools for validating the xml file. Later in the xml, I include the Cobbler template elements. After the wingen trigger is executed, we can also validate the autoinstall xml for each Cobbler profile.

In addition, autoinstall xml has a namespace and the file can be validated using any standard external xml tools.
```xml
<? xml version = "1.0" encoding = "utf-8"?>
<unattend xmlns = "urn: schemas-microsoft-com: unattend">
```
The trigger code does not provide for the validation of the resulting file. It seems to me that it would be redundant. The purpose of the trigger is to generate autoinstall xml and other files according to the supplied template.

This is what Microsoft tools validation looks like:
validate-answer-file-windows-10

**Comment by tpw56j** [comment]
> Please fix the linter. The CI is our most important way of ensuring that Cobbler works. Since the linter is already complaining, 

@SchoolGuy Okay, a little later, I'll clean up the code and prepare other necessary files to isolate the trigger into a separate package.

**Comment by SchoolGuy** [comment]
@tpw56j Why do you want to separate the trigger into a dedicated package? From my side this is not needed. Fixing the linter (& Travis Tests should pass) would be enough for me to merge this PR.

**Comment by tpw56j** [comment]
@SchoolGuy
I am not very good at code design, so I do not insist on separating the trigger into a dedicated package.

I think there are two arguments for separating into a dedicated package:
- The trigger has dependencies that differ from those in the main package: python3-pefile python3-hivex wimlib-utils.
- Not everyone has a need to install Windows, much less prepare the necessary files in this way.

If these arguments are irrelevant, then I'll leave the trigger in the main code.

**Comment by SchoolGuy** [comment]
@tpw56j We have the same scenario with the ldap auth module. Thus leave it. We have it documented what dependencies are needed, if someone wants to use the trigger. We improve the behavior with dependencies at a later point if we have time.

**Comment by tpw56j** [comment]
@SchoolGuy Ok, then I’m going to clean up the code.

**Comment by SchoolGuy** [comment]
Note: Codacy may fail in its current state because the tools there are too strict for the current project state.

**Comment by SchoolGuy** [comment]
I restarted Travis because I don't understand why the tests are failing. Thus I restarted them. If they fail again then I guess this trigger changes something in Cobbler which causes them to break.

**Comment by SchoolGuy** [comment]
@tpw56j When you think this is ready to be merged, remove the WIP prefix and I will merge after my final review.

**Review by SchoolGuy — APPROVED** [review]
CI passes, Linter is green. This is ready from my eyes.

**Comment by SchoolGuy** [comment]
Let's merge it. Thanks a lot again @tpw56j! You will be mentioned in the next release for sure!

**Pull request merged**
